File: algorithms/python/24Game/24Game.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/24-game/
# Author: Miao Zhang
# Date:   2021-02-28

class Solution:
    def judgePoint24(self, nums: List[int]) -> bool:
        self.TARGET = 24
        self.EPS = 1e-6
        self.ADD, self.MULTIPLY, self.SUBTRACT, self.DIVIDE = 0, 1, 2, 3
        return self.solve(nums)

    def solve(self, nums: List[float]) -> bool:
        if not nums: return False
        if len(nums) == 1: return abs(nums[0] - self.TARGET) < self.EPS
        for i, x in enumerate(nums):
            for j, y in enumerate(nums):
                if i == j: continue
                t = []
                for k, z in enumerate(nums):
                    if k != i and k != j:
                        t.append(z)
                for op in range(4):
                    if op < 2 and i > j: continue
                    if op == self.ADD:
                        t.append(x + y)
                    elif op == self.MULTIPLY:
                        t.append(x * y)
                    elif op == self.SUBTRACT:
                        t.append(x - y)
                    elif op == self.DIVIDE:
                        if abs(y) < self.EPS: continue
                        t.append(x / y)
                    if self.solve(t): return True
                    t.pop()
        return False

File: algorithms/cpp/24Game/24Game.cpp
// Source: https://leetcode.com/problems/24-game/
// Author: Miao Zhang
// Date:   2021-03-01

class Solution {
public:
    double EPS = 1e-6;
    int TARGET = 24;
    int ADD = 0, MULTIPLY = 1, SUBTRACT = 2, DIVIDE = 3;

    bool judgePoint24(vector<int>& nums) {
        vector<double> num(nums.begin(), nums.end());
        return solve(num);
    }

private:
    bool solve(vector<double>& nums) {
        if (nums.size() == 0) return false;
        if (nums.size() == 1) return abs(nums[0] - TARGET) < EPS;
        for (int i = 0; i < nums.size(); i++) {
            for (int j = 0; j < nums.size(); j++) {
                if (i == j) continue;
                vector<double> t;
                for (int k = 0; k < nums.size(); k++) {
                    if (k != i && k != j) t.push_back(nums[k]);
                }
                for (int op = 0; op < 4; op++) {
                    if (op < 2 && i > j) continue;
                    if (op == ADD) {
                        t.emplace_back(nums[i] + nums[j]);
                    } else if (op == MULTIPLY) {
                        t.emplace_back(nums[i] * nums[j]);
                    } else if (op == SUBTRACT) {
                        t.emplace_back(nums[i] - nums[j]);
                    } else if (op == DIVIDE) {
                        if (fabs(nums[j]) < EPS) {
                            continue;
                        }
                        t.emplace_back(nums[i] / nums[j]);
                    }
                    if (solve(t)) return true;
                    t.pop_back();
                }
            }
        }
        return false;
    }
};
